---
title: "Module_11_maf_tools"
author: "Nick Chevalier"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 1) Introduction
Mutation Annotation Format (MAF) is being widely accepted and used to store somatic variants detected. The Cancer Genome Atlas Project has sequenced over 30 different cancers with sample size of each cancer type being over 200. 
The Resulting data consisting of somatic variants are stored in the form of Mutation Annotation Format. This package attempts to summarize, analyze, annotate and visualize MAF files in an efficient manner from either TCGA sources or any in-house studies as long as the data is in MAF format.

## 2) Generating MAF
If have VCF files or simple tabular files, then can use vcf2maf utility
This will annotate VCFs, prioritize transcripts, and generates an MAF. Recent updates to gatk has also enabled funcotator to genrate MAF files.
If using ANNOVAR for variant annotations, maftools has a handy function annovarToMaf for converting tabular annovar outputs to MAF.


## 6) Reading and Summarizing maf files

### 6.1 Input files
an MAF file - can be gz compressed. Required.
an optional but recommended clinical data associated with each sample/Tumor_Sample_Barcode in MAF.
an optional copy number data if available. Can be GISTIC output or a custom table containing sample names, gene names and copy-number status (Amp or Del).

### 6.2 Reading MAF files
```{r 6.2_reading-maf-files}
library("maftools")
#path to TCGA LAML MAF file
laml.maf = system.file('extdata', 'tcga_laml.maf.gz', package = 'maftools') 
#clinical information containing survival information and histology. This is optional
laml.clin = system.file('extdata', 'tcga_laml_annot.tsv', package = 'maftools') 

laml = read.maf(maf = laml.maf, clinicalData = laml.clin)
```
read.maf function reads MAF files, summarizes it in various ways and stores it as an MAF object. Even though MAF file is alone enough, it is recommended to provide annotations associated with samples in MAF. One can also integrate copy number data if available.

Note that by default, Variants with High/Moderate consequences are considered as non-synonymous. You change this behavior with the argument vc_nonSyn in read.maf.

### 6.3 MAF Object
``` {r 6.3_maf-object}
# Show the basic summary of MAF file.
laml

#Shows sample summry.
getSampleSummary(laml)
#Shows gene summary.
getGeneSummary(laml)
#shows clinical data associated with samples
getClinicalData(laml)
#Shows all fields in MAF
getFields(laml)
#Writes maf summary to an output file with basename laml.
write.mafSummary(maf = laml, basename = 'laml')
```
Summarized MAF file is stored as an MAF object. MAF object contains main maf file, summarized data and any associated sample annotations.
There are accessor methods to access the useful slots from MAF object.

## 7 Visualization

### 7.1 Plotting MAF summary
``` {r 7.1_plotting-MAF-summary}
plotmafSummary(maf = laml, rmOutlier = TRUE, addStat = 'median', dashboard = TRUE, titvRaw = FALSE)

```
Use plotmafSummary to plot the summary of the maf file, which displays number of variants in each sample as a stacked barplot and variant types as a boxplot summarized by Variant_Classification.

### 7.2 Onccoplots
#### 7.2.1 Drawing oncoplots
``` {r 7.2.1_Drawing-oncoplots}
# oncoplot for top ten mutated genes
oncoplot(maf = laml, top = 10)
```
Using oncoplot function to create waterfall plots and better represent maf file. 
Variants annotated as Multi_Hit are those genes which are mutated more than once in the same sample

### 7.3 Transition and Transversions
``` {r 7.3_transition-and-transversions}
laml.titv = titv(maf = laml, plot = FALSE, useSyn = TRUE)
#plot titv summary
plotTiTv(res = laml.titv)
```
titv function classifies SNPs into transitions and transversions and returns a list of summarized tables in various ways. 
-Transitions are the changing of 2 purines or of 2 pyrimidines (Base changes of similar shape):
purines: A->G or G<->A (2 rings)
pyrimidines: C->T or T->C (1 ring)
-Transversions are interchanges of purine for pyrimidine:
purine change to pyrimidine: G->C/T and A->C/T or
pyrimidine change to purine: C->A/G and T->A/G 
There are 2x transversion options compared to transitions. However, transition mutations are generated at higher frequency than transversions. 
Transitions are less likely to result in AA substitution (due to wobble point) - usually silent substitutions in SNPs (single nucleotide polymorphisms)

### 7.4 Lollipop plots for amino acid changes
Lollipop plots
uses AA change information in the maf file.
MAF files have no clear guidelines on naming the field for AA changes. -different studies have different field (column) names for AA changes
By default, lollipopPlot looks for column AAChange, and if it is not found in the MAF, it prints all available fields with warning message.

Below example, MAF file contains AA changes under a field/column name 'Protein_Change'. Manually specify this using AACol
#### 7.4.1 MAF as an input
``` {r 7.4.1_lollipop-plots-maf-as-input}
#lollipop plot for DNMT3A, which is one of the most frequent mutated gene in Leukemia.
lollipopPlot(
  maf = laml,
  gene = 'DNMT3A',
  AACol = 'Protein_Change',
  showMutationRate = TRUE,
  labelPos = 882
)
```

#### 7.4.2 Custom data as an input
``` {r 7.4.2_custom-data-as-input}
#example data
my_data = data.frame(pos = sample.int(912, 15, replace = TRUE), count = sample.int(30, 15, replace = TRUE))
my_data
lollipopPlot(data = my_data, gene = "DNMT3A")

# lollipopPlot documentation example using the TCGA laml data:
#laml.maf <- system.file("extdata", "tcga_laml.maf.gz", package = "maftools")
#laml <- read.maf(maf = laml.maf)
#lollipopPlot(maf = laml, gene = 'KIT', AACol = 'Protein_Change')

# plot general protein domains
plotProtein(gene = "TP53", refSeqID = "NM_000546")

```

### 7.5 Rainfall plots
Cancer genomes, especially solid tumors are characterized by genomic loci with localized hyper-mutations. Such hyper mutated genomic regions can be visualized by plotting INTER variant distance on a linear genomic scale - rainfall
detectChangePoints is set to TRUE, rainfall plot also highlights regions where potential changes in inter-event distances are located.
``` {r 7.5_rainfall-plots}
brca <- system.file("extdata", "brca.maf.gz", package = "maftools")
brca = read.maf(maf = brca, verbose = FALSE)

rainfallPlot(maf = brca, detectChangePoints = TRUE, pointSize = 0.4)
```
"Lataegis" are defined as those genomic segments containing six or more consecutive mutations with an average inter-mutation distanc eof less than or equal to 100 bp to 5

### 7.6 Compare Mutation load against TCGA cohorts
tsgaCompare uses mutation load from TCGA MC3 for comparing mutation burden against 33 TCGA cohorts. Plot generates is similar to the one described in Alexandrob et al.
``` {r 7.6_compare-mutation-load-against-TCGA-cohorts}
laml.mutload = tcgaCompare(maf = laml, cohortName = 'Example=LAML', logscale = TRUE, capture_size = 50)
```

### 7.7 Plotting VAF
This function plots Variant Allele Frequencies as a boxplot which quickly helps estimate the clonal status of top mutated genes (clonal genes usually have mean allele frequency around ~50% assuming pure sample)
``` {r 7.7_plotting-vaf}
plotVaf(maf = laml, vafCol = 'i_TumorVAF_WU')
```

# Skipping part 8 


### 9 Analysis
#### 9.1 Somatic Interactions
Mutually exclusive or co-occurring set of genes can be detected using somaticInteractions function. Performs pair-wise Fisher's Exact test to detect significant pair of genes
``` {r 9.1_somatic-interactions}
#exclusive/co-occurrence event analysis on top 10 mutated genes.
somaticInteractions(maf = laml, top = 25, pvalue = c(0.05, 0.1))
```

#### 9.2 Detecting cancer driver genes based on positional clustering
maftools function oncodrive - identifies cancer genes (driver) from a MAF.
Is based on an algorithm oncodriveCLUST which was originally implemented in Pythin. Concept based on the fact that most of variants in cancer causing genes are enriched at few specific loci (hot-spots). Takes advantage of positions to identify cancer gene drivers: 
``` {r 9.2_oncodrive-detecting-cancer-drivers}
# run oncodrive with protein change and z-score to find what genes are enriched
laml.sig = oncodrive(maf = laml, AACol = 'Protein_Change', minMut = 5, pvalMethod = 'zscore')
head(laml.sig)

# plot the results as scatter plot
plotOncodrive(res = laml.sig, fdrCutOff = 0.1, useFraction = TRUE, labelSize = 0.5)
```

#### 9.3 Adding and summarizing pfam domains
pfamDomains adds pfam domain information to AA changes and summarizes AA changes according to the domains affected. 
Able to find what domain in cancer cohort is most frequently affected.
``` {r 9.3_pfamDomain}
laml.pfam = pfamDomains(maf = laml, AACol = "Protein_Change", top = 10)
#Protein summary (Printing first 7 columns)
laml.pfam$proteinSummary[,1:7, with = FALSE]

#Domain summary (printing first 3 columns)
laml.pfam$domainSummary[,1:3, with = FALSE]
```

#### 9.4 Survival Analysis
mafSurvive performs survival analysis and draws kaplan meier curve by grouping samples based on mutation status of user defined gene(s) or maunally provided samples. 
input data must contain Tumor_Sample_Barcode (make sure they match in the MAF files), binary event (1/0), and time of event.

This example annotation data already contains survival information. If we have survival data stored in separate table provide via argument clinicalData
##### 9.4.1 Mutation in any given genes
``` {r 9.4.1_mafSurvival}
#Survival analysis based on grouping of DNMT3A mutation status
laml.maf <- system.file("extdata", "tcga_laml.maf.gz", package = "maftools")
laml.clin <- system.file("extdata", "tcga_laml_annot.tsv", package = "maftools")
laml <- read.maf(maf = laml.maf,  clinicalData = laml.clin)
mafSurvival(maf = laml, genes = 'DNMT3A', time = 'days_to_last_followup', Status = 'Overall_Survival_Status', isTCGA = TRUE)

```

##### 9.4.2 Predict gene sets associated with survival
``` {r 9.4.2_gene-sets-survival}
#Using top 20 mutated genes to identify a set of genes (of size 2) to predict poor prognostic groups
prog_geneset = survGroup(maf = laml, top = 20, geneSetSize = 2, time = "days_to_last_followup", Status = "Overall_Survival_Status", verbose = FALSE)
# combination (N=2) of genes associated with poor survival (P<0.05).
print(prog_geneset)

# Draw KM curves
mafSurvGroup(maf = laml, geneSet = c("DNMT3A", "FLT3"), time = "days_to_last_followup", Status = "Overall_Survival_Status")
```

#### 9.5 Comparing 2 Cohorts
Cancers differ from each other in terms of their mutation pattern. We compare 2 different cohorts to detect differentially mutated genes. 
Ex: Madan et. al showed patients with relapsed APL (acute promyelocytic Leukemia) tends to have mutations in PML and RARA genes, which were absent during primary stage of disease. Difference between 2 cohorts (primary v relapsed APL) can be detected using mafCompare 
mafCompare - performs fisher test on all genes between 2 cohorts to detect differentially mutated genes
``` {r 9.5_mafCompare}
#Primary APL MAF
primary.apl = system.file("extdata", "APL_primary.maf.gz", package = "maftools")
primary.apl = read.maf(maf = primary.apl)

# Relapse APL MAF
relapse.apl = system.file("extdata", "APL_relapse.maf.gz", package = "maftools")
relapse.apl = read.maf(maf = relapse.apl)

#Considering only genes which are mutated in at-least 5 samples in 1 cohort to avoid bias due to genes mutated in single sample
pt.vs.rt <- mafCompare(m1 = primary.apl, m2 = relapse.apl, m1Name = 'Primary', m2Name = 'Relapse', minMut = 5)
print(pt.vs.rt)
```

#### 9.5.1 Forest Plots
Use forest plots to visualize the results of primary vs relapse APL for PML and RARA
``` {r 9.5.1_forest-plots}
# visualize the results of primary vs relapse APL for PML and RARA
forestPlot(mafCompareRes = pt.vs.rt, pVal = 0.1)
```

#### 9.5.2 Co-onco plots
alternative way to display results is by plotting 2 oncoplots side by side. 
coOncoplot function takes 2 mafs and plots them side by side for better comparison
``` {r 9.5.2_co-onco-plots}
genes = c("PML", "RARA", "RUNX1", "ARID1B", "FLT3")
coOncoplot(m1 = primary.apl, m2 = relapse.apl, m1Name = 'PrimaryAPL', m2Name = 'ReplapseAPL', genes = genes, removeNonMutated = TRUE)
```

#### 9.5.3 Co-bar plots
``` {r 9.5.3_co-bar-plots}
coBarplot(m1 = primary.apl, m2 = relapse.apl, m1Name = "Primary", m2Name = "Relapse")
```

#### 9.5.4 Lollipop plot-2
Along with plots showing cohort-wise differences. Its also possible to show gene-wise differences with lollipopPlot2 function
```{r 9.5.4_lollipop-plot-2}
lollipopPlot2(m1 = primary.apl, m2 = relapse.apl, gene = "PML", AACol1 = "amino_acid_change", AACol2 = "amino_acid_change", m1_name = "Primary", m2_name = "Relapse")
```

#### 9.6 Clinical Enrichment Analysis
clinicalEnrichment is a function that takes clinical feature associated with the samples and performs enrichment analysis. 
It performs various groupwise and pairwise comparisons to identify enriched mutations for every category within a clinical feature. 
Below ex identifies mutations associated with FAB_classification
``` {r 9.6_clinical-enrichment}
fab.ce = clinicalEnrichment(maf = laml, clinicalFeature = 'FAB_classification')

#Results are returned as a list. Significant associations p-value < 0.05
fab.ce$groupwise_comparision[p_value < 0.05]
```
Results shows IDH1 muts are enriched in M1 subtype of leukemia compared to the rest of the cohort. 
DNMT3A is in M5, RUNX1 is in M0, etc. (is well known and function recaptures the results). 
Next we plot the results:
``` {r 9.6_plot-enricment}
plotEnrichmentResults(enrich_res = fab.ce, pVal = 0.05, geneFontSize = 0.5, annoFontSize = 0.6)
```

#### 9.7 Drug-Gene Interactions
drugInteractions function checks for drug-gene interactions and gene druggability information compiled from "Drug Gene interactions" database
``` {r 9.7_drug-interactions}
dgi = drugInteractions(maf = laml, fontSize = 0.75)
```
Plot shows potential druggable gene categories along with top 5 genes involved in them. 
Can also extract information on drug-gene interactions
Ex: below is the results for known/reported drugs to interact with DNMT3A
``` {r 9.7_drug-interactions-DNMT3A}
dnmt3a.dgi = drugInteractions(genes = "DNMT3A", drugs = TRUE)
# Printing selected columns
dnmt3a.dgi[,.(Gene, interaction_types, drug_name, drug_claim_name)]
```


#### 9.8 Oncogenic Signaling Pathways
pathways function checks for enrichment of known Oncogenic Signaling Pathways from TCGA cohorts.
``` {r 9.8_oncogenic-signling}
# There is no fucntion pathways or plotPathways with maftools. Now it is OncogenicPathways and plotOncogenicPathways
pws = pathways(maf = laml, plotType = 'treemap')
plotPathways(maf = laml, pathlist = pws)
```
